component/attr_cache/attr_cache.go

```diff
@@ -501,46 +478,63 @@ func (ac *AttrCache) RenameDir(options internal.RenameDirOptions) error {
+		ac.cacheLock.Lock()
+		defer ac.cacheLock.Unlock()
+
+		// TLDR: Dst is guaranteed to be non-existent or empty.
+		// Note: We do not need to invalidate children of Dst due to the logic in our FUSE connector, see comments there,
+		// but it is always safer to double check than not.
+		ac.invalidateDirectory(options.Dst)
+
+		// if attr_cache is tracking directories, validate this rename
 		if ac.cacheDirs {
-			// renameCachedDirectory may cache the parent directory
-			// so lock the cache for writing
-			ac.cacheLock.Lock()
-			defer ac.cacheLock.Unlock()
-			err = ac.renameCachedDirectory(options.Src, options.Dst, currentTime)
+			// First, check if the destination directory already exists
+			if ac.pathExistsInCache(options.Dst) {
+				return os.ErrExist
+			}
+		}
+
+		// get the source directory
+		srcItem, getErr := ac.cacheMap.get(options.Src)
+		if getErr != nil || !srcItem.exists() {
+			log.Err("AttrCache::RenameDir : %s source not found", options.Src)
+			if ac.cacheDirs {
+				return syscall.ENOENT
+			}
 		} else {
-			ac.cacheLock.RLock()
-			defer ac.cacheLock.RUnlock()
-			ac.deleteDirectory(options.Src, currentTime)
-			// TLDR: Dst is guaranteed to be non-existent or empty.
-			// Note: We do not need to invalidate children of Dst due to the logic in our FUSE connector, see comments there,
-			// but it is always safer to double check than not.
-			ac.invalidateDirectory(options.Dst)
+			// move everything over
+			srcDir := internal.TruncateDirName(options.Src)
+			dstDir := internal.TruncateDirName(options.Dst)
+			ac.moveAttrCachedItem(srcItem, srcDir, dstDir, currentTime)
 		}
 	}
 
 	return err
 }
 
-// CreateFile: Mark the file invalid
+// CreateFile: Cache a new entry for the file
 func (ac *AttrCache) CreateFile(options internal.CreateFileOptions) (*handlemap.Handle, error) {
 	log.Trace("AttrCache::CreateFile : %s", options.Name)
 	h, err := ac.NextComponent().CreateFile(options)
 
 	if err == nil {
+		currentTime := time.Now()
 		// TODO: the cache locks are used incorrectly here
 		// They routinely lock the cache for reading, but then write to it
 		ac.cacheLock.RLock()
 		defer ac.cacheLock.RUnlock()
 		if ac.cacheDirs {
 			// record that the parent directory tree contains at least one object
-			ac.markAncestorsInCloud(getParentDir(options.Name), time.Now())
+			ac.markAncestorsInCloud(getParentDir(options.Name), currentTime)
 		}
-		// TODO: we assume that the OS will call GetAttr after this.
-		// 		if it doesn't, will invalidating this entry cause problems?
-		toBeInvalid, getErr := ac.cacheMap.get(options.Name)
-		if getErr == nil {
-			toBeInvalid.invalidate()
+		newFileEntry, getErr := ac.cacheMap.get(options.Name)
+		if getErr != nil || !newFileEntry.exists() {
+			// add new entry
+			newFileAttr := internal.CreateObjAttr(options.Name, 0, currentTime)
+			newFileEntry = ac.cacheMap.insert(newFileAttr, true, currentTime)
 		}
+		newFileEntry.setSize(0, currentTime)
+		newFileEntry.setMode(options.Mode)
```

Comment:
```suggestion
		// add new entry
		newFileAttr := internal.CreateObjAttr(options.Name, 0, currentTime)
		newFileEntry = ac.cacheMap.insert(newFileAttr, true, currentTime)
		newFileEntry.setMode(options.Mode)
```
I think this is a simpler and better implementation.
No re-using a left-over entry which should have been invalidated, and since `insert` replaces old entries, we don't even need to look up the old entry to invalidate it.

Reply:
Do we need to call setSize here, or was that unnecessary.

Reply:
The size is set by `CreateObjAttr` (the `0` parameter).